tardis/plasma/properties/nlte_rate_equation_solver.py

```diff
@@ -0,0 +1,367 @@
+        lte_rate_matrix_block += n_e_matrix
+        lte_rate_matrix_block[-1, :] = 1.0
+        return lte_rate_matrix_block
+
+    @staticmethod
+    def prepare_phi(phi):
+        """
+        Makes sure that phi does not have any 0 entries.
+        """
+        phi[phi == 0.0] = 1.0e-10 * phi[phi > 0.0].min().min()
+        return phi
+
+    @staticmethod
+    def recomb_matrix(recomb_coefficients, atomic_number, ion_number):
+        """Constructs a recombination rate matrix from the recombination rates.
+
+        Parameters
+        ----------
+        recomb_coefficients : DataFrame
+            Recombination coefficients.
+        atomic_number : int64
+            Current atomic number. Used for the dimension of a square matrix.
+        ion_number : int64
+            Current ion number. Used for returning the correct row.
+
+        Returns
+        -------
+        numpy.ndarray
+        """
+        offdiag = np.zeros(atomic_number)
+        index = recomb_coefficients.index
+        for i in index:
+            offdiag[i] = recomb_coefficients.loc[i]
+        diag = np.hstack([np.zeros(1), -offdiag])
+        return (np.diag(diag) + np.diag(offdiag, k=1))[ion_number, :]
+
+    @staticmethod
+    def ion_matrix(ion_coefficients, atomic_number, ion_number):
+        """Constructs an ionization rate matrix from the ionization rates.
+
+        Parameters
+        ----------
+        ion_coefficients : DataFrame
+            Recombination coefficients.
+        atomic_number : int64
+            Current atomic number. Used for the dimension of a square matrix.
+        ion_number : int64
+            Current ion number. Used for returning the correct row.
+
+        Returns
+        -------
+        numpy.ndarray
+        """
+        offdiag = np.zeros(atomic_number)
+        index = ion_coefficients.index
+        for i in index:
+            offdiag[i] = ion_coefficients.loc[i]
+        diag = np.hstack([-offdiag, np.zeros(1)])
+        return (np.diag(diag) + np.diag(offdiag, k=-1))[ion_number, :]
+
+    @staticmethod
+    def prepare_charge_conservation_row(atomic_numbers):
+        """Prepares the last row of the rate_matrix. This row corresponds to the charge density equation."""
+        charge_conservation_row = []
+        for atomic_number in atomic_numbers:
+            charge_conservation_row.append(np.arange(0.0, atomic_number + 1))
+        charge_conservation_row = np.hstack([*charge_conservation_row, -1])
+        # TODO needs to be modified for use in nlte_excitation
+        return charge_conservation_row
+
+    @staticmethod
+    def prepare_ion_recomb_coefficients_nlte_ion(
+        gamma,
+        alpha_sp,
+        alpha_stim,
+        coll_ion_coeff,
+        coll_recomb_coeff,
+        partition_function,
+        levels,
+        level_boltzmann_factor,
+    ):
+        """
+        Prepares the ionization and recombination coefficients by grouping them for ion numbers.
```

Comment:
Not a complete docstring, but you can probably copy the input variable descriptions from another function.

Reply:
Done